### ISSUE TITLE
Asynchronous validation?

### ISSUE THREAD

**Issue by grahamb** [olahol/react-tagsinput#11]
Hi!

First, thanks for this repo; it's been very easy to integrate this component into a form I'm building.

I was wondering if you had any suggestions for how to handle async validation? I'm using react-tagsinput to let user enter a list of usernames. I need to validate those username against a REST API. It doesn't look like it would work out-of-the-box; `addTag` calls `validate` and expects a synchronous response. There's also the question of what to do with an invalid tag when an async response comes back.

**Comment by olahol** [comment]
Currently there is no way to do asynchronous validation of tags, but I am adding support for it in version 1.3.0 which should be coming up tomorrow.

**Comment by grahamb** [comment]
👍 awesome! 

— Graham

**Comment by olahol** [comment]
I have made `validate` asynchronous if it takes two arguments `tag` and function `cb`. You can see an example in [example/async.html](https://github.com/olahol/react-tagsinput/blob/master/example/async.html).

**Comment by grahamb** [comment]
I've been playing with this all morning, and I can't get async validations to work. No matter what I try, the `done` parameter is coming up as `undefined`.

Here's my implementation:

```javascript
import React from 'react/addons';
import TagsInput from 'react-tagsinput'

import api from 'utils/api';

const SpaceInitialUsersField = React.createClass({

  mixins: [React.addons.LinkedStateMixin],

  getInitialState() {
    return {
      tags: []
    };
  },


  validate: function (tag, done) {
    var unique = this.state.tags.indexOf(tag) === -1;
    if (!unique) { return done(false); }

    api.validate_sfu_username(tag, (result) => {
      const ok = result.status === 'ok';
      done(ok);
    });
  },

  focusInput(event) {
    this.refs.space_initial_users.getDOMNode().querySelector('input').focus();
  },

  render: function() {
    return (
      <div onClick={this.focusInput} className="ic-Form-control">
        <label htmlFor="space_initial_users" className="ic-Label">Initial Users</label>
        <div className="SFU-tagsinput-wrapper">
          <TagsInput
            name="space_initial_users"
            ref="space_initial_users"
            valueLink={this.linkState('tags')}
            placeholder="e.g. kipling@sfu.ca"
            classNamespace="SFU"
            addKeys={[9, 13, 32, 188]} // tab, return, space, comma
            removeKeys={[]}
            validate={this.validate}
          />
        </div>
      </div>
    );
  }

});

export default SpaceInitialUsersField;
```

The `example/async.html` demo works fine, so I'm guessing I'm missing something in my implementation.

**Comment by olahol** [comment]
What version of babel are you using? I was unable to replicate your bug, but I have added a prop `validateAsync` in version 1.3.1 that is explicitly asynchronous, try using it and check back with me with the results :pray:

**Comment by grahamb** [comment]
I'm using webpack with babel-loader 5.0.0, which pulls in babel-core 5.1.3.

I created a simplified test case (https://github.com/grahamb/react-tagsinput-test) with the same webpack build and it works there, so I'm not sure where I'm going wrong. I'll try pulling in 1.3.1 and get back to you.

**Comment by grahamb** [comment]
OK, using `validateAsync` works. I'm still going to dig around and see if I can figure out why `validate` isn't working, but at least I've got a workaround. Thanks!

**Comment by olahol** [comment]
Since the async behavior of `validate` depends on `function.length` there could be problems if there are any functions wrapping `validate`, maybe using the `arguments` pseudo-array to pass arguments down to `validate`, this would mangle 'functions.length` making it 0 and thus not async.

**Comment by grahamb** [comment]
Nothing like that going on, unless webpack is mucking it up.

I'm trying to figure out how to cancel validation on an empty tag. I don't want it to fail validation. Consider this case:

```
  validate: function (tag, done) {
    const unique = this.state.tags.indexOf(tag) === -1;

    if (!unique) { return done(false); }

    if (tag !== '') {
      api.validate_sfu_username(tag, (result) => {
        done(result.valid_user === true);
      });
    }
  }
```

![untitled](https://cloud.githubusercontent.com/assets/146111/7380783/b7b0343c-edb2-11e4-9857-8dd7ca812c0b.gif)

I'm not returning anything from my `validate` function when the tag is empty, so the `validating` state never gets cleared. If I return done(true) when I have an empty tag, an empty tag gets added to my state (as expected). That's not what I want, though; it should be ignored and the next input should be focused.

**Comment by olahol** [comment]
`onBeforeTags` was removed because its use case is covered by `transform` and `onChange`.

It's already possible to do what you want by using `value` and `onChange` instead of `valueLink`, it's a bit ugly but you can allow empty strings in `validate` and then check whether the new tag (the last element of the argument to `onChange`) is empty and if it is then simply don't update the `tags` state.

But this is not needed as I have added a small fix (version 1.3.2) that ignores falsey values (like empty string) from `transform`.
